JShell: Running "Execute Java Shell" from Maven project fails

matthiasblaesing opened the pull request:
When trying to run "Execute Java Shell" for a non-modular maven project, this fails because the VM can't be started.

The problem is, that NetBeans tries to start the VM with a module setup, but the project does not provide one (it is non-modular).

Please see individual commits for detailed explanation. 

Closes: #6692

matthiasblaesing commented:
@sdedic the core of this change is in [1a41b91](https://github.com/apache/netbeans/pull/6905/commits/1a41b91367d9a511b5d4f99a068098b11c0a17ef). You touched that code in 2017 in a commit titled  "#270831: quoted possible spaces in agent specification / other arguments generated by JShell", I don't understand why in that changeset `isModularProject` was modified.

Could you please check if this update make sense to you?

sdedic commented on java/jshell.support/src/org/netbeans/modules/jshell/editor/CompletionFilter.java:
```diff
@@ -55,6 +61,29 @@
+        MethodHandle getCharactersBuilder = null;
+        Lookup lookup = MethodHandles.lookup();
+        try {
+            getTypeBuilder = lookup.findVirtual(DocTrees.class, "getType", MethodType.methodType(TypeMirror.class, DocTreePath.class));
```
I'd suggest a more prominent LOG message, as `getType()` method will NPE if this `findVirtual` fails.
I am not very familiar with shimming different javac versions - @jlahoda should definitely review this.

matthiasblaesing replied:
Raised log level to warning - in a normal situation `DocTrees` is loaded from nb-javac and thus the two methods are present and can be resolved here. If it fails the methods are not present and thus can't be called in the first place, so I think this should be safe.

sdedic commented on java/jshell.support/src/org/netbeans/modules/jshell/tool/JShellTool.java:
```diff
@@ -1691,50 +1754,60 @@ boolean cmdClasspath(String arg) {
+        try {
+            if (arg.isEmpty()) {
+                debug = !debug;
+                MH_SET_DEBUG_FLAGS.invoke(state, debug ? DBG_GEN : 0);
```
maybe guard against missing method ?

matthiasblaesing replied:
Thanks for bringing this up. I added a null guard, that gives info to the user. While checking/testing this, I noticed, that my testing was tainted by modifications to the `netbeans.conf`. The package `jdk.internal.jshell.debug` needs to be exported to the unnamed module, that necessary change was also added.

sdedic commented on java/jshell.support/src/org/netbeans/modules/jshell/support/ShellSession.java:
```diff
@@ -944,8 +945,34 @@ public void notifyClosed(JShellEnvironment env, boolean remote) {
+            try {
+                startException.process().getInputStream().transferTo(baos);
+            } catch (IOException ex) {}
+            processOutput.append("Standard output:\n");
```
L10n as it goes to UI ?

matthiasblaesing replied:
Valid point - updated via 25ee1a49b5838031bd0d52294d66628ca43512a8

lahodaj commented:
(I'll (try to) take a look tomorrow, sorry.)

lahodaj reviewed (approved):
Makes sense to me.

matthiasblaesing commented:
@sdedic @lahodaj thank you both for review.

I addressed the review comments, squashed the updates into their base counterparts. github actions is still green, so I'll merge.
